Ku/partialsum

unalmis opened the pull request.

unalmis commented on desc/equilibrium/coords.py:
```diff
@@ -503,6 +503,132 @@ def fixup(x, *args):
+
+    TODO(#1243) Do proper partial summation once the DESC
+    DESC basis are improved to store the padded tensor product basis.
+    https://github.com/PlasmaControl/DESC/issues/1243#issuecomment-3131182128.
```
@PlasmaControl/desc-dev in #1508 header it says "figure out how to do FourierZernike". Basically until the basis is padded https://github.com/PlasmaControl/DESC/issues/1243#issuecomment-3131182128 there is no efficient implementation without loops.

unalmis commented:
As this pull request made no changes to the benchmarks that show speed improvement, we can conclude those benchmarks are useless and misleading.

dpanici commented:
Do you have a script we can profile on master and here to see the speed improvement you imply?

> As this pull request made no changes to the benchmarks that show speed improvement (see previous benchmark run), we can conclude those benchmarks are useless and misleading.

unalmis commented:
> Do you have a script we can profile on master and here to see the speed improvement you imply?
> 
> 
> 
> > As this pull request made no changes to the benchmarks that show speed improvement (see previous benchmark run), we can conclude those benchmarks are useless and misleading.
> 
> 
> 
> 

no i mean click to github actions benchmark from the previous run. there will be a bunch of green speed improvement according to DESC benchmarks. however those are  all wrong because this PR doesn't change any of the code those benchmarks run

unalmis commented:
It's the same issue on the laplace solver PR and your recent https://github.com/PlasmaControl/DESC/pull/1828#issuecomment-3138046359. the benchmarks are just noise

dpanici commented:
weird, it really should not be making htem green if the change is within the std dev... I agree they are quite noisy but I guess their use isif we completely tank a given function it should come up from the noise. Would be nice if we could run them on specific architectures with dedicated resources though.

unalmis commented on desc/equilibrium/equilibrium.py:
```diff
@@ -1339,55 +1345,21 @@ def _get_rtz_grid(
-            full_output=full_output,
-            **kwargs,
+        return _map_clebsch_coordinates(
+            iota, alpha, zeta, L_lmn, L, theta0, period, tol, maxiter, **kwargs
```
this is to avoid circular imports

dpanici commented on desc/objectives/_fast_ion.py:
```diff
@@ -305,50 +314,57 @@ def compute(self, params, constants=None):
+        Y_B = self._hyperparam.pop("Y_B")
+        num_transit = self._hyperparam.pop("num_transit")
+        num_quad = self._hyperparam.pop("num_quad")
+        del self._constants["X"]
```
why the deletion?

unalmis replied:
it's just not used so I wanted to delete it.

f0uriest commented on desc/equilibrium/equilibrium.py:
```diff
@@ -1339,55 +1345,21 @@ def _get_rtz_grid(
-    def compute_theta_coords(
-        self, flux_coords, L_lmn=None, tol=1e-6, maxiter=20, full_output=False, **kwargs
+    @staticmethod
+    def _map_clebsch_coordinates(
```
does this need to be a method here? I'd vote for just using the function directly where needed

unalmis replied:
It has to be an object method to avoid circular imports.

f0uriest commented on desc/equilibrium/coords.py:
```diff
@@ -376,58 +375,100 @@ def fixup(x, *args):
+        ResolutionWarning,
+        msg="High frequency lambda modes will be truncated in coordinate mapping.",
+    )
+    lmbda_minus_iota_omega = L.transform(L_lmn)
```
why is this lambda-iota*omega and not just lambda?

unalmis replied:
The root-finding from desc coords to clebsch coords involves both lambda and omega

f0uriest commented on desc/equilibrium/coords.py:
```diff
@@ -376,58 +375,100 @@ def fixup(x, *args):
-        Shape (k, ).
+        Spectral coefficients for λ.
+    L : Transform
+        Transform for λ built on coordinates [ρ, θ, ζ].
```
what is theta here?

also I'd vote to name this `L_transform` or something to be more explicit and not confuse with resolution parameter `L`

unalmis replied:
DESC theta. This is currently a private method, eventually once 1243 is done, the passed in parameter will be changed to just the spectral coefficients (no transform) and we can then do proper partial summation

unalmis replied:
I'll rename to what you want on `ku/nufft`. It's not worth running GitHub CI again for a name change for private function

f0uriest commented on desc/equilibrium/coords.py:
```diff
@@ -436,71 +477,52 @@ def _map_clebsch_coordinates(
-                **kwargs,
-            )
+    @partial(jnp.vectorize, signature="(),(),(m)->()")
+    def vecroot(theta0, alpha, c_m):
```
I think you'll still have the issue @YigitElma mentioned about recompilation here. The fix would be to make rootfun/jacfun/vecroot global (private) functions rather than define them locally within this function

unalmis replied:
In `ku/nufft` I jitted the outer function as well.

f0uriest commented on desc/equilibrium/coords.py:
```diff
@@ -436,71 +477,52 @@ def _map_clebsch_coordinates(
-        lmbda = A @ L_lmn
-        alpha_k = theta + lmbda - iota * zeta
-        return _fixup_residual(alpha_k - alpha, period).squeeze()
+    def rootfun(theta, alpha, c_m):
```
is alpha here really theta_PEST?

unalmis replied:
I hope the new labels clarify.

f0uriest commented on desc/equilibrium/coords.py:
```diff
@@ -436,71 +477,52 @@ def _map_clebsch_coordinates(
-    return out
+    c_m, modes = _partial_sum(L, L_lmn, None, None, iota)
+    c_m = c_m[:, jnp.newaxis]
+    alpha = alpha[:, jnp.newaxis] + iota[:, jnp.newaxis, jnp.newaxis] * zeta
```
is this really theta_PEST?

unalmis replied:
No, it is not supposed to be theta_PeST

unalmis commented on desc/equilibrium/coords.py:
```diff
@@ -123,18 +131,9 @@ def map_coordinates(  # noqa: C901
-                **kwargs,
-            )
+            rho, alpha, zeta = coords.T
+            coords = jnp.column_stack([rho, alpha + iota * zeta, zeta])
```
This assumes `zeta=Phi`. I have added comment to address this.

unalmis commented on desc/equilibrium/equilibrium.py:
```diff
@@ -1334,61 +1341,67 @@ def _get_rtz_grid(
+            **kwargs,
+        )
+
+    def _compute_iota_under_jit(self, rho, params=None, profiles=None, **kwargs):
```
this is used in `map_coordinates` and `Bounce2D.compute_theta`

unalmis commented:
I had to pull changes from `ku/nuffts` to address Rory's comments. So the codecov lowered by 1%. It's fine.

rahulgaur104 commented on desc/equilibrium/coords.py:
```diff
@@ -115,33 +124,27 @@ def map_coordinates(  # noqa: C901
+                iota = eq._compute_iota_under_jit(coords, params, profiles, **kwargs)
+            rho, alpha, zeta = coords.T
+            omega = 0  # TODO(#568)
+            coords = jnp.column_stack([rho, alpha + iota * (zeta + omega), zeta])
```
If zeta is the generalized toroidal angle, don't we assume zeta = phi + Omega where phi is the cylindrical toroidal angle?
So shouldn't theta_PEST = alpha + iota * (zeta - omega)?

rahulgaur104 commented on desc/equilibrium/coords.py:
```diff
@@ -376,131 +382,147 @@ def fixup(x, *args):
+        msg="High frequency lambda modes will be truncated in coordinate mapping.",
+    )
+    lmbda_minus_iota_omega = lmbda.transform(L_lmn)
+    lmbda_minus_iota_omega = (
```
So meshgrid reshape is reshaping from rtz to rzt?
What is the division and multiplication by 2 doing? Can you explain?

rahulgaur104 commented on desc/equilibrium/coords.py:
```diff
@@ -376,131 +382,147 @@ def fixup(x, *args):
-    guess=None,
-    period=np.inf,
+    lmbda,
+    theta0=None,
```
theta0 is a very bad variable name. theta0 is dangerously close to theta_0 that people in the gyrokinetics communit use in flux tube codes to define the location of vanishing integrated magnetic shear. 
Why did you change it from guess? I strongly recomment you change it back. guess is great because it immediately tells you what it is. theta0 requires user to figure out what is happening in the rest of the code.

rahulgaur104 replied:
As a rule of thumb, please don't make unnecessary changes to variable names.

unalmis replied:
1. This is a new function, so nothing is an unnecessary change because nothing is a change; it is new.[^1]
2. `theta0` is the initial guess in the root finding for `theta`. I will change it to `guess` for you. Whatever the name is, the public documentation next to the parameter `theta0: jnp.ndarray : Optional initial guess for the computational coordinates.` is there. So no one has to read code to figure out what it should be.

[^1]: The function `_map_clebsch_coordinates` is private and not used anywhere in DESC. If there is external code that was using this function, we are not responsible for breaking API convention.

rahulgaur104 commented on desc/equilibrium/coords.py:
```diff
@@ -376,131 +382,147 @@ def fixup(x, *args):
-    out : ndarray
-        Shape (k, 3).
-        DESC computational coordinates [ρ, θ, ζ].
-    info : tuple
```
Why did you remove info? This is not a good coding practice. This is basically passive encryption.

rahulgaur104 replied:
What happens if root finding fails for some reason and I want to debug it.

rahulgaur104 replied:
If the residual is calculated elsewhere, ignore this comment.

unalmis replied:
This is a new function, see https://github.com/PlasmaControl/DESC/pull/1826#discussion_r2275331172, so I disagree with the statement about coding practice. Whatever code users have that uses root finding, nothing has changed. They can still get their `info` tuple.

In this new function, I have not added functionality to return auxiliary information about the root finding simple because it is impossible - functions that decorated with `jnp.vectorize` must return arrays.

unalmis replied:
I am not convinced that this root finding can fail by the way.

rahulgaur104 commented on desc/integrals/_bounce_utils.py:
```diff
@@ -79,8 +79,8 @@ def bounce_points(pitch_inv, knots, B, dB_dz, num_well=None):
 
     """
     intersect = polyroot_vec(
```
This routine find the intersection of lines of constant pitch with the magnetic field B, i.e., the bounce points.
B is an array with a shape rho, alpha, 1/lambda (inv_pitch), something, num_wells.
Is that right?

unalmis replied:
I put the shape of all my stuff in the public documentation string (docstring for short) because I find it helpful. Here is the definition of B and its shape in that docstring. 

```python
B : jnp.ndarray
        Shape (..., N - 1, B.shape[-1]).
        Polynomial coefficients of the spline of B in local power basis.
        Last axis enumerates the coefficients of power series. Second to
        last axis enumerates the polynomials that compose a particular spline.
```
When you see `Shape (..., N - 1, B.shape[-1])`, in Python, that means the code is making the following contract with you, the developer:

The last two axes need to have that shape. All the leading axes ``...`` don't matter, the code is agonstic to them. If there are leading axes, the code will simply perform whatever it does to the last two axes in a vectorized manner. That is the same contract that `numpy` makes with you. Numpy calls this contract its "broadcasting conventions".

Less experienced developers will mess up this broadcasting contract, and instead may promise broadcasting on the trailing axes of the form `Shape (N - 1, B.shape[-1], ...)`. You should NEVER do that. If you broadcast on the trailing axes like that you break from numpy convention, and the user-facing code will have to have a million transposes and reshapes to get their stuff working with what you wrote.

rahulgaur104 commented on desc/integrals/bounce_integral.py:
```diff
@@ -145,7 +145,7 @@ class Bounce2D(Bounce):
     -----
     Magnetic field line with label α, defined by B = ∇ψ × ∇α, is determined from
-      α : ρ, θ, ζ ↦ θ + λ(ρ,θ,ζ) − ι(ρ) [ζ + ω(ρ,θ,ζ)]
+      α : ρ, θ, ζ ↦ θ + Λ(ρ,θ,ζ) − ι(ρ) [ζ + ω(ρ,θ,ζ)]
```
I am confused again by this equation. Specifically the + sign before omega.
Is the convention zeta = phi + omega or zeta = phi - omega, where phi is the cylindrical angle and zeta is the generalized angle.

unalmis replied:
The latter. One can control +f search desc/compute/_core.py for `name="phi"` and `name="theta_PEST"` to see the definitions.

rahulgaur104 reviewed (commented):
Mostly minor comments. Addressing them will help future development/ developers.

unalmis commented:
I addressed the comments. I think at developer meeting you all wanted me to add more of my pull request comments to code. Please make a new pull request and add whichever of my comments into the code that you would like to add.
